tls/s2n_ktls_io.c

```diff
@@ -268,6 +269,46 @@ S2N_RESULT s2n_ktls_recvmsg(void *io_context, uint8_t *record_type, uint8_t *buf
     return S2N_RESULT_OK;
 }
 
+/* ktls does not support updating keys, so we should kill the connection when the
+ * key encryption limit is reached. We could get the actual sequence number from
+ * the kernel with getsockopt, but that requires a surprisingly expensive syscall.
+ *
+ * Instead, we track the estimated sequence number and enforce the limit based
+ * on that estimate.
+ */
+static S2N_RESULT s2n_ktls_enforce_estimated_record_limit(
+        struct s2n_connection *conn, size_t bytes_written)
+{
+    RESULT_ENSURE_REF(conn);
+    if (bytes_written == 0) {
+        return S2N_RESULT_OK;
+    }
+
+    struct s2n_blob seq_num = { 0 };
+    RESULT_GUARD(s2n_connection_get_sequence_number(conn, conn->mode, &seq_num));
+
+    /* The RFC states the encryption limits in terms of "full-size records" sent.
+     * We can estimate the number of "full-sized records" sent by assuming that
+     * all records are full-sized.
+     */
+    do {
+        RESULT_GUARD_POSIX(s2n_increment_sequence_number(&seq_num));
+        bytes_written -= MIN(bytes_written, S2N_TLS_MAXIMUM_FRAGMENT_LENGTH);
+    } while (bytes_written > 0);
+
+    uint64_t next_seq_num = 0;
+    RESULT_GUARD_POSIX(s2n_sequence_number_to_uint64(&seq_num, &next_seq_num));
+    RESULT_ENSURE_GT(next_seq_num, 0);
+    uint64_t curr_seq_num = next_seq_num - 1;
```

Comment:
I think you're right, but it'd definitely hard to reason about. This might assume we need enough leeway to send a KeyUpdate, which isn't the case here?

I rewrote it so that it's hopefully clearer, and added a test.